File: day3/day3_2.cpp
#include <iostream>
#include <fstream>
#include <regex>


int main()
{
    std::ifstream file("input.txt");
    std::string input;
    std::string line;
    int result = 0;
    while (std::getline(file, line))
    {
        input += line;
    }

    std::regex pattern(R"(mul\([0-9]+,[0-9]+\)|do\(\)|don't\(\))");
    std::smatch match;

    std::regex_search(input, match, pattern);
    
    auto words_begin = std::sregex_iterator(input.begin(), input.end(), pattern);
    auto words_end = std::sregex_iterator();
    bool enabled = true;
    for (auto i = words_begin; i != words_end; ++i)
    {
        std::smatch match = *i;
        std::string match_str = match.str();
        if (match_str == "do()")
        {
            enabled = true;
        }
        else if (match_str == "don't()")
        {
            enabled = false;
        }
        else
        {
            if (enabled)
            {
                int a,b;
                sscanf(match_str.c_str(), "mul(%d,%d)", &a, &b);
                result += a * b;
            }
        }
    }
    std::cout << result << std::endl;

}

File: day3/day3_2.py
import re

def multiply(s: str):
    s = s.strip("mul()").split(",")
    return int(s[0]) * int(s[1])
    
with open("input.txt") as file:
    text = "".join(file.readlines())

valid_strings = re.findall("mul\([0-9]+,[0-9]+\)|do\(\)|don't\(\)", text)

ans = 0
enabled = True
for s in valid_strings:
    if s[0] == "m" and enabled:
        ans += multiply(s)
    elif s == "do()":
        enabled = True
    else:
        enabled = False
print(ans)
    
    
